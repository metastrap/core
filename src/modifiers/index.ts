export * as nextModifier from './next';
