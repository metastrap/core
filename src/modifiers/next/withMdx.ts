import type { ParseResult } from '@babel/parser';
import type { File } from '@babel/types';
import * as t from '@babel/types';
import {
  TAstMap, TObject,
  TROAstMap,
} from 'types';

export const mdxFiles = [
  'mdx-components.jsx',
  'app/blog/page.mdx',
];

export default function withMdx(
  baseAstMap: TAstMap,
  fullAstMap: TROAstMap,
) {
  // eslint-disable-next-line no-debugger
  debugger;
  /* package.json */
  const pkg = baseAstMap.get('package.json') as TObject;
  pkg.devDependencies = {
    ...pkg.devDependencies || {},
    '@next/mdx': 'latest',
  };

  nextConfigTransform(
    baseAstMap.get('next.config.js') as ParseResult<File>,
  );

  mdxFiles.forEach((file) => {
    baseAstMap.set(file, fullAstMap.get(file) || '');
  });
}

function nextConfigTransform(nextConfig: ParseResult<File>) {
  const { body } = nextConfig.program;

  /* find the line where 'nextConfig' is declared */
  const nextConfigDeclaration = body.find(
    (node) => node.type === 'VariableDeclaration'
      && node.declarations.some(
        (declaration) => (declaration.id as t.Identifier).name === 'nextConfig',
      ),
  ) as t.VariableDeclaration;

  /* assign 'pageExtensions' in place of empty object */
  nextConfigDeclaration.declarations[0].init = t.objectExpression([
    t.objectProperty(
      t.identifier('pageExtensions'),
      t.arrayExpression([
        t.stringLiteral('js'),
        t.stringLiteral('jsx'),
        t.stringLiteral('mdx'),
      ]),
    ),
    t.objectProperty(
      t.identifier('experimental'),
      t.objectExpression(
        [
          t.objectProperty(
            t.identifier('mdxRs'),
            t.booleanLiteral(true),
          ),
        ],
      ),
    ),
  ]);

  /* require `@next/mdx` with init */
  body.unshift(
    t.variableDeclaration('const', [
      t.variableDeclarator(
        t.identifier('withMDX'),
        t.callExpression(
          t.callExpression(
            t.identifier('require'),
            [
              t.stringLiteral('@next/mdx'),
            ],
          ),
          [],
        ),
      ),
    ]),
  );

  /* find `module.exports` */
  const moduleExports = body.find(
    (node) => node.type === 'ExpressionStatement'
      && node.expression.type === 'AssignmentExpression'
      && node.expression.left.type === 'MemberExpression'
      && (node.expression.left.object as t.Identifier).name === 'module'
      && (node.expression.left.property as t.Identifier).name === 'exports',
  ) as t.ExpressionStatement;

  /* replace `module.exports = withMDX(nextConfig)` */
  (moduleExports.expression as t.AssignmentExpression).right = t.callExpression(
    t.identifier('withMDX'),
    [
      t.identifier('nextConfig'),
    ],
  );
}
