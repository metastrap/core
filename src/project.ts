import JSZip from 'jszip';

import type { TAstValue } from 'types/index';

export function pack(data: Map<string, string>): JSZip {
  const zip = new JSZip();
  data.forEach((content, file) => {
    zip.file(file, content);
  });
  return zip;
}

export function setBaseProject(
  fullProject: Map<string, TAstValue>,
  baseFiles: string[],
): Map<string, TAstValue> {
  const baseProject = new Map<string, TAstValue>();
  fullProject.forEach((content, file) => {
    if (baseFiles.includes(file)) {
      baseProject.set(file, content);
    }
  });
  return baseProject;
}
