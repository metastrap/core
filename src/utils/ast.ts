/* browser */
import { ParseResult, ParserPlugin, parse } from '@babel/parser';
import generate from '@babel/generator';

import type { File } from '@babel/types';

import { TAstMap, TAstValue } from 'types';
import logger from './logger';

enum EFileFeature {
  JAVASCRIPT = 1,
  TYPESCRIPT = 2,
  JSX = 4,
  JSON = 8,
}

export function convertTextToAst(
  fileContentMap: Map<string, string>,
): TAstMap {
  const resultMap = new Map<string, TAstValue>();

  fileContentMap.forEach((content, file) => {
    let featureFlag = 0;
    const extname = file.split('.').pop() || '';
    const babelParsePlugins: ParserPlugin[] = ['classProperties'];

    let result: TAstValue = content;
    switch (extname) {
      case 'js':
        featureFlag |= EFileFeature.JAVASCRIPT;
        break;
      case 'ts':
        featureFlag |= EFileFeature.TYPESCRIPT;
        break;
      case 'jsx':
        featureFlag |= EFileFeature.JSX;
        break;
      case 'tsx':
        featureFlag |= EFileFeature.TYPESCRIPT | EFileFeature.JSX;
        break;
      case 'json':
        try {
          result = JSON.parse(content);
        } catch (e) {
          logger((e as Error).message);
        }
    }

    if (featureFlag & EFileFeature.TYPESCRIPT) babelParsePlugins.push('typescript');
    if (featureFlag & EFileFeature.JSX) babelParsePlugins.push('jsx');

    if (featureFlag & (EFileFeature.JAVASCRIPT | EFileFeature.TYPESCRIPT | EFileFeature.JSX)) {
      try {
        result = parse(content, {
          sourceType: 'module',
          plugins: babelParsePlugins,
        });
      } catch (e) {
        logger((e as Error).message);
        result = content;
      }
    }
    resultMap.set(file, result);
  });
  return resultMap;
}

export function print(ast: Map<string, TAstValue>): Map<string, string> {
  const resultMap = new Map<string, string>();
  ast.forEach((astValue, file) => {
    if (typeof astValue === 'string') {
      resultMap.set(file, astValue);
      return;
    }
    if (astValue.program) {
      resultMap.set(file, generate(astValue as ParseResult<File>).code);
      return;
    }
    if (typeof astValue === 'object') {
      try {
        resultMap.set(file, JSON.stringify(astValue, null, 2));
      } catch (e) {
        resultMap.set(file, astValue.toString?.() || '');
      }
    }
  });
  return resultMap;
}
