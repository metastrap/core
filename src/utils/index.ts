export * as astUtils from './ast';
export * as logger from './logger';
export * as read from './read';
