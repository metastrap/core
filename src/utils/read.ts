import JSZip from 'jszip';

export default async function readZipToMap(
  zip: JSZip,
): Promise<Map<string, string>> {
  const files = Object.keys(zip.files);

  const maybeContentList = files.map(async (file) => {
    if (!zip.files[file].dir) {
      const contents = await zip.files[file].async('text');
      return [
        file,
        contents || '',
      ] as [string, string];
    }
    /* directories will return null */
    return null;
  });

  const contentList = (
    await Promise.all(maybeContentList)
  ).filter((file) => !!file);

  return new Map(contentList);
}
