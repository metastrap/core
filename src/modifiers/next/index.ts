import { withTailwindcss } from './withTailwindcss';
import withMdx from './withMdx';

/* browser */
export const baseFiles = [
  '.gitignore',
  'app/layout.jsx',
  'app/page.jsx',
  'jsconfig.json',
  'next.config.js',
  'package.json',
];

export const modifiers = {
  withTailwindcss,
  withMdx,
};
