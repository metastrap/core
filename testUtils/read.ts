// eslint-disable-next-line import/no-extraneous-dependencies
import fg from 'fast-glob';
import path from 'node:path';
import fs from 'node:fs/promises';
import { EFrameworks } from '@/constants/enum';
import { TAstValue } from 'types/index';
import { convertTextToAst } from '@/utils/ast';
import { setBaseProject } from '@/project';
import { baseFiles } from '@/modifiers/next';

export async function readFilesToMap(
  dir: string,
): Promise<Map<string, string>> {
  const files = await fg([`${dir}/**/*`], { dot: true });
  return new Map(
    await Promise.all(
      files.map(async (file) => {
        const contents = await fs.readFile(file, 'utf8');
        return [
          file.replace(`${dir}/`, ''),
          contents || '',
        ] as [string, string];
      }),
    ),
  );
}

export async function readFixture(framework: EFrameworks): Promise<Map<string, string>> {
  return await readFilesToMap(
    path.join(process.cwd(), 'fixtures', framework.toString()),
  );
}

export async function astFromFixture(): Promise<[
  Map<string, TAstValue>,
  Map<string, TAstValue>
]> {
  const fullProject = convertTextToAst(
    await readFilesToMap(
      path.join(process.cwd(), 'fixtures', 'next'),
    ),
  );
  return [
    fullProject,
    setBaseProject(fullProject, baseFiles),
  ];
}
