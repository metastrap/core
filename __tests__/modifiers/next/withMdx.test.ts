import path from 'node:path';
import { TAstValue } from 'types/index';
import { baseFiles } from '@/modifiers/next';
import { convertTextToAst } from '@/utils/ast';
import { setBaseProject } from '@/project';
import { readFilesToMap } from 'testUtils/read';
import withMdx, { mdxFiles } from '@/modifiers/next/withMdx';

let baseAsTree: Map<string, TAstValue>; let
  fullProject: Map<string, TAstValue>;

describe('modifiers/next/withMDx', () => {
  beforeEach(async () => {
    fullProject = convertTextToAst(
      await readFilesToMap(
        path.join(process.cwd(), 'fixtures', 'next'),
      ),
    );
    baseAsTree = setBaseProject(fullProject, baseFiles);
  });

  test('withMdx', () => {
    withMdx(baseAsTree, fullProject);
    expect(
      mdxFiles.map((file) => typeof baseAsTree.get(file)),
    ).toStrictEqual([
      'object', 'string',
    ]);
    expect(baseAsTree.get('next.config.js'))
      .toMatchSnapshot();
  });
});
