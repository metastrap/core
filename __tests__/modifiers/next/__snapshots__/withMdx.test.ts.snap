// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`modifiers/next/withMDx withMdx 1`] = `
Node {
  "comments": [],
  "end": 53,
  "errors": [],
  "loc": SourceLocation {
    "end": Position {
      "column": 0,
      "index": 53,
      "line": 4,
    },
    "filename": undefined,
    "identifierName": undefined,
    "start": Position {
      "column": 0,
      "index": 0,
      "line": 1,
    },
  },
  "program": Node {
    "body": [
      {
        "declarations": [
          {
            "id": {
              "name": "withMDX",
              "type": "Identifier",
            },
            "init": {
              "arguments": [],
              "callee": {
                "arguments": [
                  {
                    "type": "StringLiteral",
                    "value": "@next/mdx",
                  },
                ],
                "callee": {
                  "name": "require",
                  "type": "Identifier",
                },
                "type": "CallExpression",
              },
              "type": "CallExpression",
            },
            "type": "VariableDeclarator",
          },
        ],
        "kind": "const",
        "type": "VariableDeclaration",
      },
      Node {
        "declarations": [
          Node {
            "end": 21,
            "id": Node {
              "end": 16,
              "loc": SourceLocation {
                "end": Position {
                  "column": 16,
                  "index": 16,
                  "line": 1,
                },
                "filename": undefined,
                "identifierName": "nextConfig",
                "start": Position {
                  "column": 6,
                  "index": 6,
                  "line": 1,
                },
              },
              "name": "nextConfig",
              "start": 6,
              "type": "Identifier",
            },
            "init": {
              "properties": [
                {
                  "computed": false,
                  "decorators": null,
                  "key": {
                    "name": "pageExtensions",
                    "type": "Identifier",
                  },
                  "shorthand": false,
                  "type": "ObjectProperty",
                  "value": {
                    "elements": [
                      {
                        "type": "StringLiteral",
                        "value": "js",
                      },
                      {
                        "type": "StringLiteral",
                        "value": "jsx",
                      },
                      {
                        "type": "StringLiteral",
                        "value": "mdx",
                      },
                    ],
                    "type": "ArrayExpression",
                  },
                },
                {
                  "computed": false,
                  "decorators": null,
                  "key": {
                    "name": "experimental",
                    "type": "Identifier",
                  },
                  "shorthand": false,
                  "type": "ObjectProperty",
                  "value": {
                    "properties": [
                      {
                        "computed": false,
                        "decorators": null,
                        "key": {
                          "name": "mdxRs",
                          "type": "Identifier",
                        },
                        "shorthand": false,
                        "type": "ObjectProperty",
                        "value": {
                          "type": "BooleanLiteral",
                          "value": true,
                        },
                      },
                    ],
                    "type": "ObjectExpression",
                  },
                },
              ],
              "type": "ObjectExpression",
            },
            "loc": SourceLocation {
              "end": Position {
                "column": 21,
                "index": 21,
                "line": 1,
              },
              "filename": undefined,
              "identifierName": undefined,
              "start": Position {
                "column": 6,
                "index": 6,
                "line": 1,
              },
            },
            "start": 6,
            "type": "VariableDeclarator",
          },
        ],
        "end": 22,
        "kind": "const",
        "loc": SourceLocation {
          "end": Position {
            "column": 22,
            "index": 22,
            "line": 1,
          },
          "filename": undefined,
          "identifierName": undefined,
          "start": Position {
            "column": 0,
            "index": 0,
            "line": 1,
          },
        },
        "start": 0,
        "type": "VariableDeclaration",
      },
      Node {
        "end": 52,
        "expression": Node {
          "end": 51,
          "left": Node {
            "computed": false,
            "end": 38,
            "loc": SourceLocation {
              "end": Position {
                "column": 14,
                "index": 38,
                "line": 3,
              },
              "filename": undefined,
              "identifierName": undefined,
              "start": Position {
                "column": 0,
                "index": 24,
                "line": 3,
              },
            },
            "object": Node {
              "end": 30,
              "loc": SourceLocation {
                "end": Position {
                  "column": 6,
                  "index": 30,
                  "line": 3,
                },
                "filename": undefined,
                "identifierName": "module",
                "start": Position {
                  "column": 0,
                  "index": 24,
                  "line": 3,
                },
              },
              "name": "module",
              "start": 24,
              "type": "Identifier",
            },
            "property": Node {
              "end": 38,
              "loc": SourceLocation {
                "end": Position {
                  "column": 14,
                  "index": 38,
                  "line": 3,
                },
                "filename": undefined,
                "identifierName": "exports",
                "start": Position {
                  "column": 7,
                  "index": 31,
                  "line": 3,
                },
              },
              "name": "exports",
              "start": 31,
              "type": "Identifier",
            },
            "start": 24,
            "type": "MemberExpression",
          },
          "loc": SourceLocation {
            "end": Position {
              "column": 27,
              "index": 51,
              "line": 3,
            },
            "filename": undefined,
            "identifierName": undefined,
            "start": Position {
              "column": 0,
              "index": 24,
              "line": 3,
            },
          },
          "operator": "=",
          "right": {
            "arguments": [
              {
                "name": "nextConfig",
                "type": "Identifier",
              },
            ],
            "callee": {
              "name": "withMDX",
              "type": "Identifier",
            },
            "type": "CallExpression",
          },
          "start": 24,
          "type": "AssignmentExpression",
        },
        "loc": SourceLocation {
          "end": Position {
            "column": 28,
            "index": 52,
            "line": 3,
          },
          "filename": undefined,
          "identifierName": undefined,
          "start": Position {
            "column": 0,
            "index": 24,
            "line": 3,
          },
        },
        "start": 24,
        "type": "ExpressionStatement",
      },
    ],
    "directives": [],
    "end": 53,
    "interpreter": null,
    "loc": SourceLocation {
      "end": Position {
        "column": 0,
        "index": 53,
        "line": 4,
      },
      "filename": undefined,
      "identifierName": undefined,
      "start": Position {
        "column": 0,
        "index": 0,
        "line": 1,
      },
    },
    "sourceType": "module",
    "start": 0,
    "type": "Program",
  },
  "start": 0,
  "type": "File",
}
`;
